File: ProblemSolve/216.cs
/***********************
 * Combination Sum III *
 ***********************/

public class Solution {
    public void Backtracking(ref List<IList<int>> ans, List<int> curCandidates, int idx, int curSum, int k, int n){   
        if(curSum == n && curCandidates.Count == k){
            ans.Add(new List<int>(curCandidates));
            return;
        }
        
        if(curCandidates.Count >= k){
            return;
        }
        
         for(int i=idx; i<=9; ++i){
            curCandidates.Add(i);
            Backtracking(ref ans, curCandidates, i + 1, curSum + i, k, n);
            curCandidates.RemoveAt(curCandidates.Count - 1);
        }
    }
    
    public IList<IList<int>> CombinationSum3(int k, int n) {
        List<IList<int>> ans = new List<IList<int>>();
        
        Backtracking(ref ans, new List<int>(), 1, 0, k, n);
        return ans;
    }
}

File: ProblemSolve/216.cpp
/***********************
 * Combination Sum III *
 ***********************/

class Solution {
public:
    void backtracking(vector<vector<int>>& ans, vector<int> curCandidates, int idx, int curSum, int k, int n){
        if(curSum == n && curCandidates.size() == k){
            ans.push_back(curCandidates);
            return;
        }
        
        if(curCandidates.size() >= k){
            return;
        }
        
        for(int i=idx; i<=9; ++i){
            curCandidates.push_back(i);
            backtracking(ans, curCandidates, i + 1, curSum + i, k, n);
            curCandidates.pop_back();
        }
        
        return;
    }
    
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ans;
        vector<int> curCandidates;
        
        backtracking(ans, curCandidates, 1, 0, k, n);
        return ans;
    }
};